src/firejail/template.c

```diff
@@ -0,0 +1,429 @@
+/*
+ * Copyright (C) 2021 Jolla Ltd.
+ * Copyright (C) 2021 Open Mobile Platform
+ *
+ * This file is part of firejail project
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along
+ * with this program; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+*/
+
+#include "firejail.h"
+#include <stdlib.h>
+#include <string.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <errno.h>
+
+#define TEMPLATE_KEY_VALUE_DELIM ":"
+#define TEMPLATE_KEY_MACRO_DELIM "$"
+#define TEMPLATE_KEY_MACRO_SUB_DELIMS "{}"
+
+typedef struct template_t {
+	char *key;
+	char *value;
+	struct template_t *next;
+} Template;
+
+typedef enum {
+	STR_CHECK_ALNUM,
+	STR_CHECK_DBUS
+} StrCheckType;
+
+Template *tmpl_list = NULL;
+
+static Template *template_new(const char *key, const char *value)
+{
+	Template *tmpl;
+
+	if (!key || !*key || !value || !*value)
+		return NULL;
+
+	tmpl = (Template*)malloc(sizeof(Template));
+
+	tmpl->key = strdup(key);
+	tmpl->value = strdup(value);
+	tmpl->next = NULL;
+
+	fprintf(stdout, "Created new template key \"%s\" value \"%s\"\n",
+								key, value);
+
+	return tmpl;
+}
+
+static void template_free(Template *tmpl)
+{
+	if (!tmpl)
+		return;
+
+	fprintf(stdout, "free %p key \"%s\" value \"%s\"\n", tmpl, tmpl->key,
+								tmpl->value);
+
+	free(tmpl->key);
+	free(tmpl->value);
+	free(tmpl);
+}
+
+static Template* template_get(const char *key, int *err)
+{
+	Template *iter;
+	int key_len;
+
+	if (!tmpl_list) {
+		*err = -ENOENT;
+		return NULL;
+	}
+
+	if (!key || !*key) {
+		*err = -EINVAL;
+		return NULL;
+	}
+
+	key_len = strlen(key);
+	iter = tmpl_list;
+	while (iter) {
+		if (!strncmp(key, iter->key, key_len))
+			return iter;
+
+		iter = iter->next;
+	}
+
+	*err = -ENOKEY;
+	return NULL;
+}
+
+static const char* template_get_value(const char *key, int *err)
+{
+	Template *tmpl = template_get(key, err);
+
+	if (!tmpl)
+		return NULL;
+
+	return tmpl->value;
+}
+
+/* Prepend template to the list */
+static int template_add(Template *tmpl)
+{
+	int err;
+
+	if (!tmpl)
+		return -EINVAL;
+
+	if (!tmpl_list) {
+		tmpl_list = tmpl;
+		return 0;
+	}
+
+	if (template_get(tmpl->key, &err))
+		return -EALREADY;
```

Comment:
Idea here is to exit firejail if the same template is defined twice as cli arg. In such case EALREADY is reported back.

Reply:
I think this works as it is intended since for me it is quite routine to check the return value and free the content if adding is not possible due to duplicate content and thus, the newly allocated must be free'd by the caller.